src/ucp/api/ucp.h

```diff
@@ -1738,6 +1755,81 @@ ucs_status_t ucp_worker_fence(ucp_worker_h worker);
  */
 ucs_status_t ucp_worker_flush(ucp_worker_h worker);
 
+typedef enum {
+    UCP_ATOMIC_POST_OP_ADD,
+    UCP_ATOMIC_POST_OP_XOR,
+    UCP_ATOMIC_POST_OP_AND,
+    UCP_ATOMIC_POST_OP_OR,
+} ucp_atomic_post_op_t;
+
+typedef enum {
+    UCP_ATOMIC_FETCH_OP_FADD,
+    UCP_ATOMIC_FETCH_OP_SWAP, /* swap with *result */
+    UCP_ATOMIC_FETCH_OP_CSWAP,
+} ucp_atomic_fetch_op_t;
+
+ucs_status_t ucp_atomic_post32(ucp_ep_h ep, ucp_atomic_post_op_t opcode, uint32_t value,
+                               uint64_t remote_addr, ucp_rkey_h rkey);
+
+ucs_status_ptr_t ucp_atomic_fetch32_nb(ucp_ep_h ep, ucp_atomic_fetch_op_t opcode,
+                                       uint32_t value, uint32_t *result,
+                                       uint64_t remote_addr, ucp_rkey_h rkey,
+                                       ucp_send_callback_t cb);
+/*
+ * shmem_iput
+ * MPI_Put
+ * multiple remote keys - need global key support in the mem map
+ *   global remote key -> pass in endpoint address
+ *   no need for remote key -> don't pass at all
+ *  if ucp_rkey_pack returns 0 -> no need to exchange the rkey
+ *  UCP_NULL_RKEY could be passed to RMA/atomics
+ */
+ucs_status_ptr_t ucp_put_nbe(ucp_ep_h ep, const void *buffer, size_t count,
+                             ucp_datatype_t datatype,
+                             uint64_t remote_addr, ucp_datatype_t remote_datatype,
+                             ucp_rkey_t rkey,
+                             ucp_send_callback_t cb);
+
```

Comment:
What about separation between local and remote completions (we can do it with separate callbacks)?